R/sensitivityCalculation.R

```diff
@@ -0,0 +1,115 @@
+#'   parameters. If not specified, the following vector will be used: c(0.1,
+#'   0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1, 2, 3, 4, 5,  6, 7, 8, 9, 10).
+#' @param pkParameters A vector of names of PK parameters for which the
+#'   sensitivities will be calculated. For a full set of available PK
+#'   parameters, run `names(ospsuite::StandardPKParameter)`. By default, only
+#'   the following parameters will be considered: `"C_max"`, `"t_max"`,
+#'   `"AUC_inf"`. If `NULL`, all available PK-parameters will be calculated.
+#' @param pkDataFilePath Path to excel file in which
+#'   PK-parameter data should be saved. If a file already exists, it will be
+#'   overwritten. Default is `NULL`, meaning the data will not be saved to a
+#'   spreadsheet.
+#'
+#' @examples
+#'
+#' simPath <- system.file("extdata", "Aciclovir.pkml", package = "ospsuite")
+#' simulation <- loadSimulation(simPath)
+#' outputPaths <- "Organism|PeripheralVenousBlood|Aciclovir|Plasma (Peripheral Venous Blood)"
+#' parameterPaths <- c(
+#'   "Aciclovir|Lipophilicity",
+#'   "Applications|IV 250mg 10min|Application_1|ProtocolSchemaItem|Dose",
+#'   "Neighborhoods|Kidney_pls_Kidney_ur|Aciclovir|Glomerular Filtration-GFR|GFR fraction"
+#' )
+#'
+#' # extract the results into a list of dataframes
+#' sensitivityCalculation(
+#'   simulation = simulation,
+#'   outputPaths = outputPaths,
+#'   parameterPaths = parameterPaths
+#' )
+#' @export
+sensitivityCalculation <- function(simulation,
+                                   outputPaths,
+                                   parameterPaths,
+                                   variationRange = c(seq(0.1, 1, by = 0.1), seq(2, 10, by = 1)),
+                                   pkParameters = c("C_max", "t_max", "AUC_inf"),
+                                   pkDataFilePath = NULL) {
+  # fail fast if specified PK parameters are not part of standard `enum`
+  validateIsIncluded(pkParameters, names(ospsuite::StandardPKParameter))
+
+  # set outputs to the provided path
+  clearOutputs(simulation)
+  addOutputs(outputPaths, simulation)
+
+  # create simulation batch for efficient calculations
+  parameters <- getAllParametersMatching(parameterPaths, simulation)
+
+  # extract a list with results
+  simulationResultsBatch <- purrr::map(
+    .x = parameters,
+    .f = ~ .extractSimulationResultsBatch(
+      simulation     = simulation,
+      parameter      = .x,
+      variationRange = variationRange
+    )
+  )
+
+  # name list with name of each parameter path
+  names(simulationResultsBatch) <- parameterPaths
+
+  # extract dataframe for PK parameters
+  pkData <- .simulationResultsBatchToPKDataFrame(simulationResultsBatch, parameterPaths)
+
+  # filter out unneeded PK parameters
+  if (!is.null(pkParameters)) {
+    pkData <- dplyr::filter(pkData, PKParameter %in% pkParameters)
+  }
+
+  # add a new `.rowid` column with unique name for each output path
+  pkData <- .addRowid(pkData)
```

Comment:
This (and line 99) is only needed if writing to excel? Then it should be in the next if-condition. In general it feels somehow strange to add a new column to a (potentially very large) data frame just to split by sheets..

Reply:
I have removed the `.addRowid()` function, and there i no longer a new column is added only to be removed later.